ENH: Add relativistic Breit-Wigner Distribution

steppi opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes gh-6414
#### What does this implement/fix?
<!--Please explain your changes.-->
This PR adds the [Relativistic Breit-Wigner Distribution](https://en.wikipedia.org/wiki/Relativistic_Breit%E2%80%93Wigner_distribution), a distribution used in high energy physics to model resonances, aka unstable particles.
#### Additional information
<!--Any additional information you think is important.-->

###### Distribution and moments

Typically, this distribution is parametrized in terms of the mass $M$ and decay width $\Gamma$ for the resonance of interest. We're using an alternative loc-scale parametrization given by @WarrenWeckesser in the discussion for gh-6414.

In the loc-scale parametrization, the pdf of the distribution is

$$
    f(E; \rho, \Gamma) = \frac{1}{\Gamma}\frac{k}{((E/\Gamma)^2 - \rho^2)^2 + \rho^2)}
$$

where

$$
    k = \frac{2\sqrt{2}\rho^2\sqrt{\rho^2 + 1}}{\pi\sqrt{\rho^2 + \rho\sqrt{\rho^2 + 1}}}
$$

The shape parameter $\rho$ is equal to $M/\Gamma$, the mass of the resonance divided by its decay width, and $\Gamma$ is the scale parameter.

There is a closed form for the CDF (now ignoring the scale parameter)

$$
F(E, \rho) = -i k\left(\tan^{-1}\left(E/c\right)/c -
                                    \tan^{-1}\left(E/\bar{c}\right)/\bar{c}\right)/2\rho
$$

where $c = \sqrt{-\rho(\rho + i)}$ and $\bar{c}$ is its complex conjugate.

This can be found by using the factorization 

$$
\left(E^2- \rho^2\right)^2 + \rho^2 = \left(E^2 - \rho^2 + \rho i\right)\left(E^2 - \rho^2- \rho i\right)
$$

expanding the expression for the pdf with partial fractions and making use of the well known integral

$$
\int \frac{\operatorname{d}x}{x^2 + c} = \frac{\tan^{-1}(x/\sqrt{c})}{\sqrt{c}}
$$

There are also closed formulas for the first and second moments that have been used. My son woke up so I can't TeX these up now, but I can add them later. Higher moments don't exist.

##### Assumption for rho

I checked some actual $M$, $\Gamma$ values for resonances (these are included in the tests), and found that typically $\rho$ is never small and can be very large. I tried to implement things to work well for large $\rho$. That's why the  formulas look a little different from what's stated above.

##### Fitting
I've overridden `rv_continuous`'s default `fit` method to better handle the case when `floc` is set. In this case we can take advantage of $\Gamma$ being approximately the interquartile range and $\rho \Gamma$ being approximately the median to get good starting values. (shifting appropriately if `floc` is set to something other than zero).

##### Complex step derivatives
The complex step derivative test (gh-4979) fails. The results are close, but fail to meet the `1e-5` tolerance. I've added this distribution to the `fails_cmplx` list. Perhaps a tweak could improve the results enough to meet the `1e-5` tolerance.

##### Naming
Provisionally, I've gone with the name `relativistic_bw`, but I'm not committed to this if anyone has a better suggestion.

##### cc
@mdhaber please take a look when you have time. @HDembinski ,if time permits, let me know if the docstring looks OK. I took some graduate physics classes many years ago, but I'm not a physicist and may have written something that's off. 

Since it's an enhancement, I'll make a post on the mailing list within a day or so.

HDembinski commented:
Hi @steppi this is looking good, but the doc string needs a pass. It is not the center-of-mass energy, M is just the mass of the particle, which is invariant = the same in all reference frames. We sometimes call it "invariant mass" to stress that, although that is a tautology, strictly speaking. You may have learned the false concept of the "moving mass" or so (not sure what it is called in english), which is taught in school, and which is not invariant. When you study proper physics you have to unlearn that.

steppi commented:
> Hi @steppi this is looking good, but the doc string needs a pass. It is not the center-of-mass energy, M is just the mass of the particle, which is invariant = the same in all reference frames. We sometimes call it "invariant mass" to stress that, although that is a tautology, strictly speaking. You may have learned the false concept of the "moving mass" or so (not sure what it is called in english), which is taught in school, and which is not invariant. When you study proper physics you have to unlearn that.

Thanks! 

Here center-of-mass energy is supposed to refer to $E$ but I can see how the wording is jumbled. $E$ isn’t even mentioned in the docstring. I cribbed the term center-of-mass energy directly from wikipedia. 

Yes, I learned about rest mass vs relativistic mass. It’s funny, this set of [lecture notes](http://www.hep.ph.ic.ac.uk/~dauncey/will/lecture02.pdf) I found when looking up the term center-of-mass energy discusses invariant mass and says that rest mass isn’t a very useful concept, but I didn’t catch notice of that until you pointed it out just now. 

Is center-of-mass energy an acceptable name for $E$ in the rbw distribution, or is Wikipedia a little off here? Is there a better term we can call it?

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -10286,6 +10286,140 @@ def _single_cdf(q, k, df):
+
+    @inherit_docstring_from(rv_continuous)
+    def fit(self, data, *args, **kwds):
+        # Override rv_continuous.fit to better handle case where floc is set.
```
Can you override `_fitstart` instead?

steppi replied:
Possibly. It depends on if we only want to support accurate fitting for the floc=0 case. Which very well may be true since this is the case that is relevant for applications. What ends up happening is that the initial point ends up biased to the floc=0 case, and because the likelihood landscape is highly multimodal, you’re liable to end up with a result close to the best floc=0 solution regardless of the data. As far as I can tell, _fitstart isn’t currently aware of the value of floc, and I’d be anxious about modifying such a crucial piece of distribution infrastructure. By overriding fit, we can make fitting work well with any value of floc. If floc isn’t set, fitting won’t work well in any case though.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -7569,3 +7569,73 @@ def test_moment_order_4():
+
+
+class TestRelativisticBW:
+    @pytest.mark.parametrize(
+        "rho,gamma", [
+            (36.545206797050334, 2.4952),  # Z0 Boson
+            (38.55107913669065, 2.085),  # W Boson,
+        ]
+    )
+    def test_pdf_integration(self, rho, gamma):
+        """Test pdf integrates to 1."""
+        res = quad(stats.relativistic_bw.pdf, 0, np.inf, args=(rho, 0, gamma))
+        assert_allclose(res[0], 1)
```
I don't think this one is in [`test_continuous_basic`](https://github.com/scipy/scipy/blob/main/scipy/stats/tests/test_continuous_basic.py), but personally I wouldn't include tests that are applicable to all distributions in distribution-specific tests.

ev-br replied:
Just a note from 10000 feet that there is a generic test of this kind: https://github.com/scipy/scipy/blob/main/scipy/stats/tests/common_tests.py#L23

mdhaber replied:
Great. So I think all of these are implemented generically.

steppi replied:
I see. I had looked at the PR for the most recently distribution added distribution (studentized range) and added the same tests that appeared there.  I noticed there was some overlap with the generic tests but ended up thinking it was good to be able to test with sets of parameters that correspond to actual particles.  The Higgs boson test identified issues with fitting I wouldn’t have found otherwise for instance but couldn’t be used in the generic tests because as far as I can tell, distr_params won’t allow using a scale parameter. 

Also, are there generic tests for fitting with floc set? This seems to be the only case where fitting rbw will actually work. 

I’ll leave it up to you two whether we should take these tests out though.

ev-br replied:
Tests from `common_tests.py` take values from https://github.com/scipy/scipy/blob/main/scipy/stats/_distr_params.py --- and these latter values end up in the default docstring examples. So all in all it sounds like the actual Higgs mass (or is it $Z^0$ boson) could/should very well end up in this list.

Whether to keep tests or not --- I'm not current on recent enhancements in stats (as is, over last several years), so am not making decisions here :-).

steppi replied:
Maybe the best solution would be to allow distr_params to take loc and scale parameters. I’m not sure how much work this would take though.

ev-br replied:
ISTM this would be reasonably easy (and it would be great indeed): 
the loop over the distributions,  https://github.com/scipy/scipy/blob/main/scipy/stats/tests/test_continuous_basic.py#L146 , takes `args`
 https://github.com/scipy/scipy/blob/main/scipy/stats/tests/test_continuous_basic.py#L142
which come straight from the list in `_distr_params.py`. So I *think* simply adding values for loc and scale to the relevant line in https://github.com/scipy/scipy/blob/main/scipy/stats/_distr_params.py#L6 should just work.

Whether the docstring generation needs a tweak for the case where loc and scale are given, I don't know, and suspect only that the only way is to step through https://github.com/scipy/scipy/blob/main/scipy/stats/_distn_infrastructure.py#L755
Most likely it'll need a slight tweak of the generic doctsting example template.

steppi replied:
Ok cool. I was getting an error when I tried that but will look into again when time permits. It seems like the error should be an easy fix

mdhaber replied:
Oops. There are some things about studentized range I'd change if I were starting it now!

You can include any of these if they're not redundant. They just looked redundant, or if they weren't, I was suggesting that they be added generically.

It's definitely fine to keep the fit tests or improve the generic ones. It would be good to run the `fit` test with user-defined `scale` because code paths for those are often parameter or data dependent.

steppi replied:
Ok, I think the best thing to do is to keep only the fit tests here and figure out why I was getting an error when trying to add loc and scale to distr_params, then use the params for the Higgs boson there.

ev-br replied:
> So I think simply adding values for loc and scale to the relevant line in https://github.com/scipy/scipy/blob/main/scipy/stats/_distr_params.py#L6 should just work.

Nope, it doesn't _just work_ :-). The reason is several tests either assume the default loc and scale (hence asserts fail), or supply `loc=0, scale=1` explicitly (hence multiple values for a loc keyword), or assume a signature of `*args, n, loc=0, scale=1` (e.g. `moment`).

So to allow custom loc and scale would need to augment the list in `_distr_param` with a dict of loc and scale keywords, , where applicable, make individual `check_` routines receive this extra `kwds` dict as a parameter (also handle the case where it is not supplied) and use pass them to `distfn` methods. 

While this is not too difficult per se, it definitely does not block this PR.

steppi replied:
Thanks for looking into that @ev-br! @mdhaber do you think it’s ok to keep some of these tests (maybe just for the most difficult case of the Higgs) until we make these changes in distr_params?

mdhaber replied:
> The Higgs boson test identified issues with fitting I wouldn’t have found otherwise

I see that adding `loc`/`scale` to the `fit` test would be useful, and I agree that we should be able to add `loc` and `scale` to `distr_params` (so that all these parameters are shown in the documentation, which would be nice here). I don't really see why `loc` and `scale` are needed for the other tests, since those parameters are handled in a generic way. Note that you can add multiple sets of parameters to `distr_params` if that's what you're going for - only one will be rendered. 

But if you prefer to include them here, I suppose they're simple enough. Please just explain as a comment in the code for posterity.

steppi replied:
> I'm happy to use custom `loc` and `scale` in the other tests, but I'm not sure why they would be needed, since those parameters are handled in a generic way.

I thought it would be good to test with actual parameter values for real particles. In that case scale will be equal to the decay width. It would have been embarrassing if some numerical issue caused problems for one of the  combinations that will actually be used. That’s what the fit tests identified for the Higgs Boson, which led to overriding fit. 

I think I’m going to get rid of all of the dist specific tests except the fitting ones though. Later I can follow up adding more combinations to distr_params after it’s able to take loc/scale parameters.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -7569,3 +7569,73 @@ def test_moment_order_4():
+            (96292.3076923077, 0.0013),  # Higgs Boson
+        ]
+    )
+    def test_pdf_against_cdf(self, rho, gamma):
```
There is a PDF-CDF consistency check in `test_continuous_basic.py`, IIRC.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -7569,3 +7569,73 @@ def test_moment_order_4():
+        ]
+    )
+    @pytest.mark.parametrize("n", [1, 2])
+    def test_moments(self, n, rho, gamma):
```
Similarly for moments, I think. Please take a look at what's already in `test_continuous_basic.py`.

mdhaber commented on doc/source/tutorial/stats/continuous_relativistic_bw.rst:
```diff
@@ -0,0 +1,32 @@
+======================================
+
+There is a single shape parameter :math:`\rho`. The support is
+:math:`0 \leq x < \infty`
```
If we add the PDF and CDF to the tutorial, I'd suggest adding the moments.

steppi replied:
Agreed

mdhaber commented on doc/source/tutorial/stats/continuous_relativistic_bw.rst:
```diff
@@ -0,0 +1,32 @@
+                                    \frac{\tan^{-1}\left(\frac{x}{\bar{c}}\right)}{\bar{c}}\right)}{2\rho}
+   \end{eqnarray*}
+
+Where
```
```suggestion
where
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -10286,6 +10286,140 @@ def _single_cdf(q, k, df):
+    %(after_notes)s
+
+    :math:`\rho = M/\Gamma` and :math:`\Gamma` is the scale parameter. If one
+    seeks to model the :math:`Z^0` boson with :math:`M \approx 91.1876 text{
```
```suggestion
    seeks to model the :math:`Z^0` boson with :math:`M \approx 91.1876 \text{
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -10286,6 +10286,140 @@ def _single_cdf(q, k, df):
+
+    %(after_notes)s
+
+    :math:`\rho = M/\Gamma` and :math:`\Gamma` is the scale parameter. If one
```
"For example, if one..." maybe?

steppi replied:
I’d dithered on this one. Let’s go with your suggestion.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -10286,6 +10286,140 @@ def _single_cdf(q, k, df):
+
+    def _stats(self, rho):
+        # Returning None from stats makes public stats use _munp.
+        return None, None, np.nan, np.nan
```
I haven't checked, but do these need to be `np.inf`, or does `np.inf` need to be `np.nan` above? Or does it end up returning the same thing either way?

steppi replied:
I had tried both and I think what happened is I ended up leaving nans because the results ended up being the same. I’ll check again though.

steppi replied:
Ah, I still need to address this one. They should be inf not nan. The raw and central moments of order greater than 2 are infinite, not undefined.

mdhaber replied:
These still report `nan`.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -10286,6 +10286,140 @@ def _single_cdf(q, k, df):
+    def _shape_info(self):
+        return [_ShapeInfo("rho", False, (0, np.inf), (False, False))]
+
+    def _pdf(self, x, rho):
```
Would implementing `_logpdf`/`_logcdf` be useful? In that case, could we exponentiate them to implement the `pdf/cdf`?

steppi replied:
I’ll look into it.

mdhaber replied:
It could be left as an enhancement, too.

mdhaber reviewed (commented):
Thanks for this @steppi! I took a quick look at documentation and tests. The main comment about tests is that generic tests that apply to all distributions are taken care of in `test_continuous_basic`. If there is a hole in those tests, please help fill it in a separate PR, but here, we could use tests that show that you are implementing the right distribution. Are there any published values you can find, or is there a reference implementation you can compare against? Could you test against a one-line implementation that you include right in the test function. This usually only makes sense if there is some simpler way to write the functions. If there isn't one, I could write implementations that you could use as independent checks. I'll probably be checking the math with SymPy in my next pass, so I could generate reference values from that.

mdhaber commented:
Yes, I agree that the relationships between the functions make it unlikely for there to be a problem as long as the generic tests pass and the PDF is correct. And the PDF is quite easily checked manually, but I think it's good to add at least some automatic check of the PDF - something to permanently tie the suite of tests to the intended distribution.

steppi commented:
> Yes, I agree that the relationships between the functions make it unlikely for there to be a problem as long as the generic tests pass and the PDF is correct. And the PDF is quite easily checked manually, but I think it's good to add at least some automatic check of the PDF - something to permanently tie the suite of tests to the intended distribution. I will suggest something when I review again.

Yeah, that makes sense. Perhaps @HDembinski knows of a table of published values.

HDembinski commented:
Sorry, I don't know of tables.

steppi commented:
I found a [reference implementation](https://github.com/root-project/root/blob/80622c45a1393cdd02990b0378a1f3c617be516f/math/mathcore/src/TMath.cxx#L452), not sure how I missed it before.

mdhaber commented:
It looks like `root` has a LGPL license. You can use reference values produced by the code, but please don't look at the code.

steppi commented:
> It looks like `root` has a LGPL license. You can use reference values produced by the code, but please don't look at the code.

Good catch. I’ve deleted the link. I’ll use it only for reference values.

steppi commented:
@mdhaber I think I've addressed all of the review comments

1. Removed tests that are redundant with common tests
2. Added an external test checking against ROOT's pdf implementation
3. Fixed errors in docstring
4. Added moments to tutorial

@HDembinski can you check if the docstring looks better now?

#### Addendum

Below is the code I used to generate the test cases using ROOT's Python bindings. I had to install a nightly build from https://root.cern/install/nightlies/  because `BreitWignerRelativistic` hasn't made it into a release yet.

```python
import numpy as np
import pandas as pd
import ROOT

from ROOT.TMath import BreitWignerRelativistic
from scipy.stats import relativistic_bw


params = (
    (36.545206797050334, 2.4952),
    (38.55107913669065, 2.085),
    (96292.3076923077, 0.0013),
)

dfs = []
for rho, gamma in params:
    X = np.linspace(0, 200, 401)
    pdf_scipy = relativistic_bw.pdf(X, rho, scale=gamma)
    pdf_root = np.fromiter(
        (BreitWignerRelativistic(x, rho * gamma, gamma) for x in X),
        dtype=float,
    )
    df = pd.DataFrame({"x": X, "pdf": pdf_root})
    df["rho"] = rho
    df["gamma"] = gamma
    dfs.append(df)
    print(np.max(np.abs(pdf_scipy - pdf_root) / np.abs(pdf_root)))
df = pd.concat(dfs)
with open("relativistic_pdf_sample_data_ROOT.npy", "wb") as f:
    np.save(f, df.values)

```

steppi commented:
@mdhaber using `np.inf`s intead of `np.nan`s for the skew and kurtosis in `stats` caused `check_skew_expect` to fail in `test_moments`. It seems like the divergence is too slow for the integrator to catch. I noticed that in the `stats`  method of `pareto`, the skew and kurtosis are given the values `nan` in cases where they technically should be infinite. I changed `relativistic_bw` back to adopting this convention. When filling these values with `nan` the end result is that `stats` only returns a tuple of length two containing only the mean and variance, which I think is reasonable.

HDembinski commented:
> @HDembinski can you check if the docstring looks better now?

Sorry, but we need still need some adjustment. You write:

```
It gives the distribution for the probability of
observing a resonance with energy :math:`E` given the rest mass :math:`M`
and decay-width :math:`\Gamma` of the resonance [1]_, where :math:`M` and
:math:`\Gamma` are expressed in natural units.
```

It now sounds like one would use the total energy $E = [m^2 + p^2]^{1/2}$ of the decaying particle in the formula, but that would not be correct. The energy is not invariant to the frame of reference. The energy to put there is actually the energy in the center-of-mass system for a two-body decay, $E_{cm} = [(E_1 + E_2)^2 - (\vec {p}_1 + \vec{p}_2)^2]^{1/2}$, like you originally wrote (sorry).

I find that misleading, like I already said, so I checked the Wikipedia and I see now that you copied the original text from there, where it also says that the energy E is the centre-of-mass energy. I then checked the bible for high energy physics, the PDG data book available at http://pdg.lbl.gov. In the 2014 edition that I have at hand. It also states that the energy in the formula is the center-of-mass energy for a two-body decay, $E_{cm} = [(E_1 + E_2)^2 - (\vec {p}_1 + \vec{p}_2)^2]^{1/2}$. 

In summary, I cannot complain that you follow these references with high reputation, but it is still misleading. The statement that the center-of-mass energy should be used is correct, because in this special system, this formula coincides with the formula to compute the invariant mass of the particle, because in the in the center-of-mass system, the parent particle that decays has velocity zero and the total energy in that system is equal to the mass of the particle.

I don't like that one should use $E_{cm}$, because it suggests to the naive reader that if you are not in the center-of-mass frame, you have to boost the energies and momenta of the particles first into this frame to be able to compute $E_{cm}$, but that is not necessary, since $E_{cm} = m$, you can use the energies and momenta $E_{1,2}$ and $\vec{p}_{1,2}$ in any reference frame! This is obviously much easier, but it is not clear when you write $E_{cm}$.

So we have two options.

1) You could go back and say that the energy E is the center-of-mass energy. I then suggest to use $E_{cm}$ as a symbol in the formulas to emphasize this instead of $E$, and I also suggest to include the formula for $E_{cm}$ for a two-body decay, so that it is clear what $E_{cm}$ is. 

2) Alternatively, and that is my preferred option, you could replace $m$ in the formulas with $m_0$ and $E$ with $m$. You then write that $m$ is the invariant mass of the particle https://en.wikipedia.org/wiki/Invariant_mass and $m_0$ is the location of the mode (is that correct?).

steppi commented:
Thanks for all the info @HDembinski! I’m going to try to assimilate everything you said and dive deeper into this. I’d like to understand all of this stuff well just for my own personal reasons. 

It looks like I also should run the tests in a 32 bit linux docker container to figure out how to set tolerances so none will fail there.

steppi commented:
@HDembinski, I think I get it now. That’s really helpful to know that $m_0$ is a parameter with no direct physical meaning. I‘ve seen it called the “characteristic mass of the resonance”. I want to go with your option 2) but also connect it with option 1) in a sentence or two to avoid confusing someone who only saw the Wikipedia definition and only has a passing knowledge of the relevant physics (like myself, though I hope to work on that.)

HDembinski commented:
That sounds perfect. The label "characteristic mass" also sounds good.

steppi commented:
I’ve changed the docstring. @HDembinski, Let me know what you think of the update.

mdhaber commented:
Thanks for reviewing the documentation @HDembinski!  Have you or would you be willing to review the correctness of the CDF/moments math and implementation?

HDembinski commented:
> Thanks for reviewing the documentation @HDembinski! Have you or would you be willing to review the correctness of the CDF/moments math and implementation?

I have limited time in the next two weeks, so I cannot help much at the moment, but I did a quick check of the formula of the CDF that is given in the description. Differentiating with sympy, I can recover the PDF.

To check the implementations, I would write some unit tests that compute the CDF via numerical intergration from the PDF, which @steppi has verified against ROOT and compare that with the implementation. Same can be done for the moments.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -7639,3 +7639,72 @@ def test_moment_order_4():
+        x, pdf = data['x'], data['pdf']
+        assert_allclose(
+            pdf, stats.relativistic_bw.pdf(x, rho, scale=gamma), rtol=rtol
+        )
```
The test above is great. Here's what I had in mind in https://github.com/scipy/scipy/pull/17505#pullrequestreview-1196914685 when I suggested testing against a simple implementation that is right in the test code. I think it's reasonable to use as a reference because it was written independently and it's very easy to check against the formulas on Wikipedia.

```suggestion
        )

    @pytest.mark.parametrize("rho, Gamma, rtol", [
              (36.545206797050334, 2.4952, 1e-14),  # Z0 Boson
              (38.55107913669065, 2.085, 1e-14),  # W Boson
              (96292.3076923077, 0.0013, 5e-13),  # Higgs Boson
          ]
      )
    def test_pdf_against_simple_implementation(self, rho, Gamma, rtol):
        # reference implementation straight from formulas on Wikipedia [1]
        def pdf(E, M, Gamma):
            gamma = np.sqrt(M**2 * (M**2 + Gamma**2))
            k = (2 * np.sqrt(2) * M * Gamma * gamma
                 / (np.pi * np.sqrt(M**2 + gamma)))
            return k / ((E**2 - M**2)**2 + M**2*Gamma**2)

        # get reasonable values at which to evaluate the CDF
        p = np.linspace(0.1, 0.9, 9)
        x = stats.relativistic_bw.ppf(p, rho, scale=Gamma)

        res = stats.relativistic_bw.pdf(x, rho, scale=Gamma)
        ref = pdf(x, rho*Gamma, Gamma)
        assert_allclose(res, ref)
```

steppi replied:
Agreed. We should include this too.

mdhaber reviewed (commented):
I'll post what I got to last night before the power went out. I'll try to look at the rest today.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -10271,6 +10271,150 @@ def _single_cdf(q, k, df):
+                2 * (1 + 1/rho**2) / (1 + np.sqrt(1 + 1/rho**2))
+            ) / np.pi
+        result = np.arctan(x/np.sqrt(-rho*(rho + 1j))) / np.sqrt(-1 - 1j/rho)
+        # For real entries of x, one can take advantage of the quantity to be
+        # subtracted being the complex conjugate of the first. This is not the
+        # case for complex entries of x with nonzero imaginary part.
```
The documentation suggests that the support is the nonegative reals. Can we remove the support for complex `x`, then?

steppi replied:
See https://github.com/scipy/scipy/pull/4979. Supporting complex numbers, (at least in a neighborhood of the real line) allows for using a complex step-wise derivative numerical trick to evaluate a finite difference derivative evaluation with one function call. It's tested [here](https://github.com/scipy/scipy/blob/78608b7c7e26908bc54fe363360d5619b54b961c/scipy/stats/tests/test_continuous_basic.py#L223-L224). The test actually fails for `relativistic_bw`, but just because the approximation can be off by a little more than the tolerance used.

mdhaber commented:
@WarrenWeckesser would you be willing to review this one? @steppi mentioned that he could take a look at gh-17587.

steppi commented on scipy/stats/_continuous_distns.py:
```diff
@@ -10271,6 +10271,150 @@ def _single_cdf(q, k, df):
+    def _shape_info(self):
+        return [_ShapeInfo("rho", False, (0, np.inf), (False, False))]
+
+    def _pdf(self, x, rho):
```
I think I should all these constants `C` or something and then include comments showing how each relates to `k`.

steppi commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -7639,3 +7639,72 @@ def test_moment_order_4():
+
+        `relativistic_bw` has special handling for these cases.
+        """
+        seed = abs(hash("relativistic_bw"))
```
Oh, good point. I just cargo culted this trick because I saw it in another test.

WarrenWeckesser commented:
I'm working on adding the relativistic Breit-Wigner distribution to [`mpsci`](https://github.com/WarrenWeckesser/mpsci).  I already added code for the PDF, CDF, inverse CDF, mean and mode (all subject to change, however), and I'm currently experimenting with different formulations for the variance and moments.  When I have a better understanding of the numerical issues involved, I'll get back to this PR (or to the code, if the PR has been merged by then).  I don't have an estimate for when that will be.

At the moment, I have only a bikeshedding comment:  I'd really like to see the names Breit and Wigner in the distribution name.  The `bw` in `relativistic_bw` seems a bit too cryptic.  FWIW, in `mpsci`, I went with `rel_breitwigner`.

steppi commented:
Oops, I should have ran the full test suite. I didn't expect anything to break.

WarrenWeckesser commented:
@steppi, just letting you know that I am still looking into this.  There are some numerical issues that I'll comment on after I have investigated a bit more.

steppi commented:
> @steppi, just letting you know that I am still looking into this. There are some numerical issues that I'll comment on after I have investigated a bit more.

Great, thanks! I appreciate that you're making such a thorough investigation of the numerical considerations. I didn't put much thought into that.

WarrenWeckesser commented on scipy/stats/_continuous_distns.py:
```diff
@@ -10289,6 +10289,158 @@ def _single_cdf(q, k, df):
+        C = np.sqrt(
+            2 * (1 + 1/rho**2) / (1 + np.sqrt(1 + 1/rho**2))
+        ) * 2 / np.pi
+        return C / (((x**2 - rho**2)/rho)**2 + 1)
```
Filter out spurious overflow warnings, and avoid the spike in the relative error when `x` is close to `rho`:
```suggestion
        with np.errstate(over='ignore'):
            return C / (((x - rho)*(x + rho)/rho)**2 + 1)
```

WarrenWeckesser reviewed (changes requested):
A few observations:

* `rel_breitwigner.pdf()` can generate a spurious overflow warning for large enough x.  E.g.
  ```
  In [3]: stats.rel_breitwigner.pdf(1e80, 5)
  [...]site-packages/scipy/stats/_continuous_distns.py:10364: RuntimeWarning: overflow encountered in square
  return C / (((x**2 - rho**2)/rho)**2 + 1)
  Out[3]: 0.0
  ```
  The computed value is fine, but the code should not generate a warning in this case. One way we can fix this is to use `np.errstate` to ignore the overflow.

  There is a reformulation of the PDF expression that would naturally avoid the overflow, and appears to improve the relative error for very large `x`, but it would mean using `lazy_where` (which affects performance), or even making the jump to a ufunc written in C.  This could be investigate in a follow-up if there is interest.  For now, the inline suggestion that I added avoids the spurious warning.  See the function `_pdf_alt()` in the gist called  `check_rel_breitwigner_pdf.py` linked below. 

* There is a spike in the relative error in `_pdf()` for `x` near `rho`.

  The following plot is generated by the script [`check_rel_breitwigner_pdf.py`](https://gist.github.com/WarrenWeckesser/aaefc2c3d6a950a5dd9068f5bfad9d4a):

  ![rel_breitwigner_pdf_rel_error](https://user-images.githubusercontent.com/321463/218318102-5606aeaa-295d-47a6-893a-59efe6b91c94.png)

  The script uses the implementation of the distribution in `mpsci` to compute the reference values.

  The first graph shows the spike in the relative error around `rho`.

  Fortunately this can be alleviated by rewriting the expression `(x**2 - rho**2)` as `(x - rho)*(x + rho)`.  There is a discussion of exactly this expression in the classic article ["What every computer scientist should know about floating-point arithmetic"](https://dl.acm.org/doi/10.1145/103162.103163).  My suggested change (inline) includes this.  The second graph in the plot shows the result.

  The third graph shows the relative error when the formulation designed specifically to handle very large `x` is added.  You can see that for `x` greater than about 1e10, the relative error is improved a bit.  The cost is adding the conditional evaluation in the function, which implies a performance hit.

* There is also spike in the relative error in `_cdf()` for `x` near `rho`.

  The script [`check_rel_breitwigner_cdf.py`](https://gist.github.com/WarrenWeckesser/0974fd7617834ef0c23e330790e8bff4) generates the following plot.

  ![rel_breitwigner_cdf_rel_error](https://user-images.githubusercontent.com/321463/218319358-becd1372-32e7-4afc-8aa2-60ade59f0eb4.png)

  I experimented a bit, but I didn't find an alternative formulation that avoids the spike in the relative error.

* Nuisance issues with the CDF:
  * Sometimes it returns a value greater than 1, e.g. `rel_breitwigner.cdf(1e8, 100)`
    returns 1.0000000000000002.  Fixing this by clipping the result should be fine.
  * For "large" rho, the CDF appears to level off at a value less than
    1 as x increases.  E.g. for large enough `x`, `rel_breitwigner.cdf(x, 1000)` returns
    0.9999999999999997.  The error is only a few ULPs, but it would be nice for the function to level off at 1.0 if it can be done without too much of a performance hit.

* I think the code in the `_cdf` method written specifically to handle complex input can be removed. I'm aware of https://github.com/scipy/scipy/pull/4979, but I don't think we should write *extra* code just to support handling complex arguments for functions where we generally expect only real inputs.  We don't need this capability in our code, so I don't think we should take on the additional burden of supporting a separate code path for complex inputs.

steppi commented:
Thanks @WarrenWeckesser! It looks like I need to read Goldberg's article again. I'll go through and make all of the other suggested changes and think about how to handle the spike near `rho` in the cdf for a little bit.

steppi commented:
@WarrenWeckesser, I think I've made all of the suggested changes. I was able to get the CDF to level off at 1.0 by fiddling with the formula a little but couldn't do anything about the spike in relative error around $\rho$. I also added support for `CensoredData` to `fit`, which was lost due to `fit` being overridden.

WarrenWeckesser commented:
@steppi, I'll update my review on Sunday (if not sooner).

steppi commented:
> @steppi, I'll update my review on Sunday (if not sooner).

Thanks!

steppi commented:
Thanks @WarrenWeckesser! I use an in editor linter to check PEP-8 and always forget it doesn't work on a file this large. I'll fix up the merge conflicts now.

steppi commented:
> . There is an asymptotic formula that could be used for large `x`, but that will need some work to figure out an appropriate order for the expansion and where to set the cutoff between using the expansion and used the simple formula `1 - cdf(x, rho)`. That can be done in a follow-up PR.

I'll try to carve out some time to follow up with this in the next month.i

steppi commented:
@mdhaber, can you give this one more look and merge if it looks good to you?

mdhaber commented:
I haven't reviewed this thoroughly, but @WarrenWeckesser has. Warren, would you be willing to merge this now that the whitespace issue has been fixed?

mdhaber commented on doc/source/tutorial/stats/continuous_rel_breitwigner.rst:
```diff
@@ -0,0 +1,38 @@
+   \begin{eqnarray*}
+       \mu & = & \frac{k}{2\rho} \left[\frac{\pi}{2} + \tan^{-1}\left(\rho\right)\right]\\
+       \mu_2 & = & \frac{k\pi}{4} \left[\frac{1 - \rho i}{\sqrt{-1 - \rho i}} + \frac{1 + \rho i}{\sqrt{-1 + \rho i}}\right]\\
+   \end{eqnarray*}
```
Could mention that the skew and kurtosis are infinite.

mdhaber reviewed (commented):
I took a quick look at the documentation and basic behavior and found some little things.

mdhaber reviewed (approved):
But assuming the math is right, I don't see anything wrong!

steppi commented:
> But assuming the math is right, I don't see anything wrong!

Thanks @mdhaber! I'm waiting for the docs to rebuild to check locally before pushing the improvements you suggested for those. I'll look into the `np.inf` vs `np.nan` issue for `_stats` a little more, and at least try to figure out why a test fails if `np.inf` is used.

mdhaber commented:
Ok, if it turns out that this is a fight against the distribution infrastructure, no need to agonize over it. We can leave these as NaNs and I can add it to my meta-issue, which I'll work on this summer.

steppi commented:
@mdhaber. I updated the docs. The failing test involves `check_skew_expect`. It looks like if the skew is not finite, then the test only allows it to be `nan`. 

```python

def check_skew_expect(distfn, arg, m, v, s, msg):
    if np.isfinite(s):
        m3e = distfn.expect(lambda x: np.power(x-m, 3), arg)
        npt.assert_almost_equal(m3e, s * np.power(v, 1.5),
                                decimal=5, err_msg=msg + ' - skew')
    else:
        npt.assert_(np.isnan(s))
```

mdhaber commented:
I'm having trouble finding a reliable source that distinguishes between moments being undefined and infinite. Can you find an example or maybe a software implementation that makes a distinction? I can see how the distinction could be useful, but I hesitate to make a change that would tread new ground when it comes to subtleties like this. If we can't find precedent, let's stick with NaN. If we can, you can change the test.

steppi commented:
> I'm having trouble finding a reliable source that distinguishes between moments being undefined and infinite. Can you find an example or maybe a software implementation that makes a distinction? I can see how the distinction could be useful, but I hesitate to make a change that would tread new ground when it comes to subtleties like this. If we can't find precedent, let's stick with NaN. If we can, you can change the test.

Let's stick with `NaN` for the time being. If we start respecting the distinction, then I think it should be done all at once for every relevant distribution, and should be brought up on the mailing list first. If I find convincing examples/source material, I'll make an issue and submit a post to mailing list.

Should I take out the $\mu_3 = \infty$, $\mu_4 = \infty$ parts of the tutorial for now to avoid confusion?

steppi commented:
@mdhaber. Wait, `._stats` doesn't return central moments, it returns `mean`, `variance`, `skew`, `kurtosis`. Skew is
$\frac{\mu_3}{\mu_2^{1.5}}$ which genuinely is undefined in this case. Something similar holds for kurtosis. So these really should be `NaN`. In that case, it's fine to have the tutorial report that the 3rd and 4th non-central moments are infinite.

mdhaber commented:
Sure, changing all at once (separately) sounds good to me. 

@WarrenWeckesser looked like this was ready for you to merge. The whitespace issue you noticed has been fixed.

WarrenWeckesser commented:
I'll take another look at this (and probably merge) on Sunday.

mdhaber commented:
Great. I'll resolve the conflicts tonight.

WarrenWeckesser reviewed (commented):
I added a question in-line about the distribution being skipped in `cases_test_fit_mle` in `test_fit.py`.  I also found that if I run the `stats` tests with `SCIPY_XSLOW` enabled, i.e.
```
SCIPY_XSLOW=1 python dev.py test -m full -j 10 -s stats
```
I get this failure:
```
============================================================= FAILURES ==============================================================
________________________ TestNumericalInversePolynomial.test_basic_all_scipy_dists[rel_breitwigner-params91] ________________________
[gw3] linux -- Python 3.10.8 /home/warren/py3.10.8/bin/python3
scipy/stats/tests/test_sampling.py:832: in test_basic_all_scipy_dists
    check_cont_samples(rng, dist, [dist.mean(), dist.var()])
        dist       = <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7f0f15c3eb30>
        distname   = 'rel_breitwigner'
        params     = (36.545206797050334,)
        rng        = <scipy.stats._unuran.unuran_wrapper.NumericalInversePolynomial object at 0x7f0f15ae5240>
        self       = <scipy.stats.tests.test_sampling.TestNumericalInversePolynomial object at 0x7f0f15fb9300>
        sup        = <numpy.testing._private.utils.suppress_warnings object at 0x7f0f15bd6f50>
scipy/stats/tests/test_sampling.py:300: in check_cont_samples
    assert_allclose(mv, mv_ex, rtol=1e-7, atol=1e-1)
        dist       = <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7f0f15c3eb30>
        mv         = (36.227478105538346, 20.51502968775722)
        mv_ex      = [36.23714553411454, 22.921329819350603]
        rng        = <scipy.stats._unuran.unuran_wrapper.NumericalInversePolynomial object at 0x7f0f15ae5240>
        rvs        = array([36.30467119, 39.08801556, 36.94529237, ..., 36.81422772,
       36.51706143, 36.33370805])
/home/warren/py3.10.8/lib/python3.10/contextlib.py:79: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Not equal to tolerance rtol=1e-07, atol=0.1
E   
E   Mismatched elements: 1 / 2 (50%)
E   Max absolute difference: 2.40630013
E   Max relative difference: 0.10498083
E    x: array([36.227478, 20.51503 ])
E    y: array([36.237146, 22.92133 ])
        args       = (<function assert_allclose.<locals>.compare at 0x7f0f15a52320>, array([36.22747811, 20.51502969]), array([36.23714553, 22.92132982]))
        func       = <function assert_array_compare at 0x7f0fe26f3400>
        kwds       = {'equal_nan': True, 'err_msg': '', 'header': 'Not equal to tolerance rtol=1e-07, atol=0.1', 'verbose': True}
        self       = <contextlib._GeneratorContextManager object at 0x7f0fe2709060>
====================================================== short test summary info ======================================================
FAILED scipy/stats/tests/test_sampling.py::TestNumericalInversePolynomial::test_basic_all_scipy_dists[rel_breitwigner-params91] - ...
=============================== 1 failed, 9104 passed, 165 skipped, 133 xfailed in 198.86s (0:03:18) ================================

```

@tirthasheshpatel, could you take a look at this?

mdhaber commented:
I don't think the test failure in `test_sampling.py` has much to do with UNURAN. 

```python3
import numpy as np
from scipy import stats
rng = np.random.default_rng(23079058616272161)
dist = stats.rel_breitwigner(36.545206797050334)
rvs = dist.rvs(100000)  # wait a long time
print(rvs.mean(), dist.mean())  # 36.24488859010438 36.23714553411454
print(rvs.var(), dist.var())  # 41.712043010121754 22.921329819350603
```

The sample variance really doesn't match the variance returned by the `var` method of the distribution. This could be because `var` is wrong, but perhaps we should not expect a finite sample variance to closely match the true variance.

A histogram of the sample produced above matches the PDF pretty closely:

```python3
import matplotlib.pyplot as plt
a, b = 30, 42.5
x = np.linspace(a, b, 200)
y = dist.pdf(x)
plt.hist(rvs, density=True, bins=np.linspace(a, b, 100))
plt.plot(x, y)
plt.title("RVS Histogram vs PDF")
plt.show()
```
![image](https://user-images.githubusercontent.com/6570539/228104227-72032ba7-ef0d-45bf-a230-b6c0bf1cacaa.png)

But `rvs` has some observations that are way out there in the tails like `1142.107459200629`, and the variance is really sensitive to them.
```python3
sorted = np.sort(rvs)
np.var(sorted)  # 41.712043010121754
np.var(sorted[:-1])  # 29.482895293876066
np.var(sorted[:-2])  # 25.174836558478603
```
This could be throwing off the sample variance.

@WarrenWeckesser With this in mind, do you expect this test to pass? I'm skipping it in the meantime.

mdhaber commented on scipy/stats/tests/test_sampling.py:
```diff
@@ -829,6 +836,8 @@ def test_basic_all_scipy_dists(self, distname, params):
             rng = NumericalInversePolynomial(dist, random_state=42)
+        if distname in skip_sample_moment_check:
+            return
         check_cont_samples(rng, dist, [dist.mean(), dist.var()])
```
```suggestion
        check_cont_samples(rng, dist, [dist.mean(), dist.var(ddof=1)])
```

mdhaber commented on scipy/stats/tests/test_sampling.py:
```diff
@@ -806,21 +806,28 @@ def test_basic(self, dist, mv_ex):
+        very_slow_dists = ['anglit', 'gausshyper', 'kappa4',
+                           'ksone', 'kstwo', 'levy_l',
+                           'levy_stable', 'studentized_range',
+                           'trapezoid', 'triang', 'vonmises']
```
I moved these from above because they are only used in `test_basic_all_scipy_dists`.

steppi commented:
Thanks @mdhaber for picking up the slack while I’ve been too busy to work on this. 

For the sample variance test. The distribution of the sample variance is fat tailed with finite mean and infinite variance, so it’s no surprise this can fail. 

<details>

If $X$ is a random variable with support on $(0, \infty)$, pdf $f(x)$ that decays like $\frac{1}{x^4}$ as $x \rightarrow \infty$, and mean $\mu$, then the random variable $Y = (X - \mu)^2$ has pdf

$$g(x) = 2\sqrt{x} f(\sqrt{x} + \mu)$$

which decays like $\frac{1}{x^{1.5}}$, implying $Y$ has finite mean but infinite variance, which implies that the sample variance

$$\frac{1}{n - 1}\sum_{i=1}^n(X_i - \mu)^2$$

for $X_i$ iid like $X$ is fat tailed with finite mean and infinite variance. 

</details>

WarrenWeckesser reviewed (commented):
One final style adjustment.  I'll commit the suggestion and merge.

WarrenWeckesser commented:
Thanks @steppi, this is a great addition. And thanks @mdhaber for the updates.
